GitUI/UserControls/RevisionGrid/Graph/LaneNodeLocator.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System.Linq;
+
+namespace GitUI.UserControls.RevisionGrid.Graph
+{
+    internal interface ILaneNodeLocator
+    {
+        (RevisionGraphRevision, bool isAtNode) FindPrevNode(int rowIndex, int lane);
+    }
+
+    internal sealed class LaneNodeLocator : ILaneNodeLocator
+    {
+        private readonly IRevisionGraphRowProvider _revisionGraphRowProvider;
+
+        public LaneNodeLocator(IRevisionGraphRowProvider revisionGraphRowProvider)
+        {
+            _revisionGraphRowProvider = revisionGraphRowProvider;
+        }
+
+        public (RevisionGraphRevision, bool isAtNode) FindPrevNode(int rowIndex, int lane)
+        {
+            if (rowIndex < 0 || lane < 0)
+            {
+                // as unlikely as it may be...
+                // don't throw, just pretend we couldn't find it
+            }
+            else
+            {
+                IRevisionGraphRow row = _revisionGraphRowProvider.GetSegmentsForRow(rowIndex);
+                if (row != null)
+                {
+                    if (row.GetCurrentRevisionLane() == lane)
+                    {
+                        return (row.Revision, isAtNode: true);
+                    }
+
+                    var segmentsForLane = row.GetSegmentsForIndex(lane);
+                    if (segmentsForLane.Count() > 0)
+                    {
+#if DEBUG
+                        var firstParent = segmentsForLane.First().Parent;
+                        foreach (var segment in segmentsForLane)
```

Comment:
I'm pretty sure this can be converted into a linq, something like 
```
if (segmentsForLane.Any(segment => segment.Parent != firstParent)) 
{
    throw(...)
}
```